mapeditor/scenelayer.cpp

```diff
@@ -0,0 +1,564 @@
+{
+}
+
+void TerrainLayer::update()
+{
+	if(!map)
+		return;
+	
+	pixmap.reset(new QPixmap(map->width * 32, map->height * 32));
+	draw(false);
+}
+
+void TerrainLayer::setDirty(const int3 & tile)
+{
+	dirty.insert(tile);
+}
+
+void TerrainLayer::draw(bool onlyDirty)
+{
+	if(!pixmap)
+		return;
+	
+	if(!map)
+		return;
+	
+	QPainter painter(pixmap.get());
+	//painter.setCompositionMode(QPainter::CompositionMode_Source);
+	
+	if(onlyDirty)
+	{
+		std::set<int3> forRedrawing(dirty), neighbours;
+		for(auto & t : dirty)
+		{
+			for(auto & tt : int3::getDirs())
+			{
+				if(map->isInTheMap(t + tt))
+					neighbours.insert(t + tt);
+			}
+		}
+		for(auto & t : neighbours)
+		{
+			for(auto & tt : int3::getDirs())
+			{
+				forRedrawing.insert(t);
+				if(map->isInTheMap(t + tt))
+					forRedrawing.insert(t + tt);
+			}
+		}
+		for(auto & t : forRedrawing)
+		{
+			handler->drawTerrainTile(painter, t.x, t.y, scene->level);
+			handler->drawRiver(painter, t.x, t.y, scene->level);
+			handler->drawRoad(painter, t.x, t.y, scene->level);
+		}
+	}
+	else
+	{
+		for(int j = 0; j < map->height; ++j)
+		{
+			for(int i = 0; i < map->width; ++i)
+			{
+				handler->drawTerrainTile(painter, i, j, scene->level);
+				handler->drawRiver(painter, i, j, scene->level);
+				handler->drawRoad(painter, i, j, scene->level);
+			}
+		}
+	}
+	
+	dirty.clear();
+	redraw();
+}
+
+ObjectsLayer::ObjectsLayer(MapSceneBase * s): AbstractLayer(s)
+{
+}
+
+void ObjectsLayer::update()
+{
+	if(!map)
+		return;
+	
+	pixmap.reset(new QPixmap(map->width * 32, map->height * 32));
+	pixmap->fill(QColor(0, 0, 0, 0));
+	draw(false);
+}
+
+void ObjectsLayer::draw(bool onlyDirty)
+{
+	if(!pixmap)
+		return;
+	
+	if(!map)
+		return;
+	
+	pixmap->fill(QColor(0, 0, 0, 0));
+	QPainter painter(pixmap.get());
+	std::set<const CGObjectInstance *> drawen;
+	
+	
+	for(int j = 0; j < map->height; ++j)
+	{
+		for(int i = 0; i < map->width; ++i)
+		{
+			handler->drawObjects(painter, i, j, scene->level);
+			/*auto & objects = main->getMapHandler()->getObjects(i, j, scene->level);
+			 for(auto & object : objects)
+			 {
+			 if(!object.obj || drawen.count(object.obj))
+			 continue;
+			 
+			 if(!onlyDirty || dirty.count(object.obj))
+			 {
+			 main->getMapHandler()->drawObject(painter, object);
+			 drawen.insert(object.obj);
+			 }
+			 }*/
+		}
+	}
+	
+	dirty.clear();
+	redraw();
+}
+
+void ObjectsLayer::setDirty(int x, int y)
+{
+	/*auto & objects = main->getMapHandler()->getObjects(x, y, scene->level);
+	for(auto & object : objects)
+	{
+		if(object.obj)
+			dirty.insert(object.obj);
+	}*/
+}
+
+void ObjectsLayer::setDirty(const CGObjectInstance * object)
+{
+}
+
+SelectionObjectsLayer::SelectionObjectsLayer(MapSceneBase * s): AbstractLayer(s), newObject(nullptr)
+{
+}
+
+void SelectionObjectsLayer::update()
+{
+	if(!map)
+		return;
+	
+	selectedObjects.clear();
+	onSelection();
+	shift = QPoint();
+	if(newObject)
+		delete newObject;
```

Comment:
deleting nullptr is safe
```suggestion
	delete newObject;
```